File: server/src/routes/courses.js
const express = require('express');
const { query, queryOne, execute } = require('../db');
const { authRequired, requireRole } = require('../middleware/auth');

const router = express.Router();

function courseOut(c) {
  return {
    id: c.id, trainer_id: c.trainer_id, title: c.title, provider: c.provider,
    skill_tags: JSON.parse(c.skill_tags || '[]'), duration_weeks: c.duration_weeks,
    level: c.level, rating: c.rating, created_at: c.created_at
  };
}

router.get('/', authRequired, async (req, res) => {
  try {
    const rows = await query('SELECT * FROM courses ORDER BY created_at DESC');
    res.json(rows.map(courseOut));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.post('/', authRequired, requireRole('trainer', 'admin'), async (req, res) => {
  try {
    const { title, provider, skill_tags, duration_weeks, level } = req.body;
    if (!title) return res.status(400).json({ error: 'title is required' });
    const result = await execute(`INSERT INTO courses (trainer_id,title,provider,skill_tags,duration_weeks,level,rating)
      VALUES ($1,$2,$3,$4,$5,$6,4.5) RETURNING id`, [
      req.user.id, title, provider || req.user.name, JSON.stringify(skill_tags || []),
      duration_weeks || 4, level || 'Beginner'
    ]);
    const c = await queryOne('SELECT * FROM courses WHERE id = $1', [result.rows[0].id]);
    res.status(201).json(courseOut(c));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Candidate enrolls
router.post('/:id/enroll', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const course = await queryOne('SELECT * FROM courses WHERE id = $1', [req.params.id]);
    if (!course) return res.status(404).json({ error: 'Course not found' });
    const existing = await queryOne('SELECT id FROM enrollments WHERE course_id=$1 AND candidate_id=$2', [req.params.id, req.user.id]);
    if (existing) return res.status(409).json({ error: 'Already enrolled' });
    await execute('INSERT INTO enrollments (course_id, candidate_id) VALUES ($1,$2)', [req.params.id, req.user.id]);
    res.status(201).json({ ok: true });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.get('/mine/enrollments', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const rows = await query(`SELECT e.*, c.title, c.provider, c.duration_weeks, c.level
      FROM enrollments e JOIN courses c ON c.id = e.course_id
      WHERE e.candidate_id = $1 ORDER BY e.created_at DESC`, [req.user.id]);
    res.json(rows);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.get('/mine/certificates', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const rows = await query(`SELECT * FROM candidate_certificates WHERE candidate_id=$1 ORDER BY issued_date DESC`, [req.user.id]);
    res.json(rows);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.get('/mine/grievances', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const rows = await query(`SELECT * FROM candidate_grievances WHERE candidate_id=$1 ORDER BY created_at DESC`, [req.user.id]);
    res.json(rows);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

router.post('/mine/grievances', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const { category, subject, description } = req.body;
    if (!subject || !description) return res.status(400).json({ error: 'subject and description are required' });
    const result = await execute(`INSERT INTO candidate_grievances (candidate_id,category,subject,description) VALUES ($1,$2,$3,$4) RETURNING id`,
      [req.user.id, category || 'Other', subject, description]);
    const row = await queryOne('SELECT * FROM candidate_grievances WHERE id=$1', [result.rows[0].id]);
    res.status(201).json(row);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Recommend courses that close the candidate's biggest skill gaps vs open jobs
router.get('/recommendations/for-me', authRequired, requireRole('candidate'), async (req, res) => {
  try {
    const me = await queryOne('SELECT skills FROM users WHERE id = $1', [req.user.id]);
    const mySkills = new Set(JSON.parse(me.skills || '[]').map(s => s.toLowerCase()));

    const jobs = await query(`SELECT required_skills FROM jobs WHERE status='open'`);
    const gapCount = {};
    for (const j of jobs) {
      for (const s of JSON.parse(j.required_skills || '[]')) {
        const key = s.toLowerCase();
        if (!mySkills.has(key)) gapCount[key] = (gapCount[key] || 0) + 1;
      }
    }
    const topGaps = Object.entries(gapCount).sort((a, b) => b[1] - a[1]).map(([s]) => s);

    const courses = await query('SELECT * FROM courses');
    const scored = courses.map(c => {
      const tags = JSON.parse(c.skill_tags || '[]').map(s => s.toLowerCase());
      const relevance = tags.filter(t => topGaps.includes(t)).length;
      return { ...courseOut(c), relevance };
    }).filter(c => c.relevance > 0).sort((a, b) => b.relevance - a.relevance);

    res.json({ topSkillGaps: topGaps.slice(0, 6), recommendedCourses: scored });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

module.exports = router;

File: server-python/routers/courses.py
import json
from fastapi import APIRouter, Depends, HTTPException
from database import query, query_one, execute, execute_returning
from auth import auth_required, require_role

router = APIRouter(prefix="/api/courses", tags=["courses"])


def _course_out(c: dict) -> dict:
    st = c.get("skill_tags")
    if isinstance(st, str):
        try: st = json.loads(st)
        except Exception: st = []
    return {
        "id": c["id"], "trainer_id": c.get("trainer_id"), "title": c["title"],
        "provider": c.get("provider"), "skill_tags": st or [],
        "duration_weeks": c.get("duration_weeks"), "level": c.get("level"),
        "rating": c.get("rating"), "created_at": c.get("created_at"),
    }


@router.get("/mine/enrollments")
async def my_enrollments(user: dict = Depends(require_role("candidate"))):
    rows = await query("""SELECT e.*, c.title, c.provider, c.duration_weeks, c.level
        FROM enrollments e JOIN courses c ON c.id=e.course_id
        WHERE e.candidate_id=$1 ORDER BY e.created_at DESC""", [user["id"]])
    return rows


@router.get("/mine/certificates")
async def my_certificates(user: dict = Depends(require_role("candidate"))):
    return await query("SELECT * FROM candidate_certificates WHERE candidate_id=$1 ORDER BY issued_date DESC", [user["id"]])


@router.get("/mine/grievances")
async def get_grievances(user: dict = Depends(require_role("candidate"))):
    return await query("SELECT * FROM candidate_grievances WHERE candidate_id=$1 ORDER BY created_at DESC", [user["id"]])


@router.post("/mine/grievances", status_code=201)
async def add_grievance(body: dict, user: dict = Depends(require_role("candidate"))):
    if not body.get("subject") or not body.get("description"):
        raise HTTPException(400, detail="subject and description are required")
    row = await execute_returning(
        "INSERT INTO candidate_grievances (candidate_id,category,subject,description) VALUES ($1,$2,$3,$4) RETURNING *",
        [user["id"], body.get("category", "Other"), body["subject"], body["description"]],
    )
    return row


@router.get("/recommendations/for-me")
async def recommendations(user: dict = Depends(require_role("candidate"))):
    me = await query_one("SELECT skills FROM users WHERE id=$1", [user["id"]])
    raw = me.get("skills") if me else None
    my_skills = {s.lower() for s in (json.loads(raw) if isinstance(raw, str) else (raw or []))}

    jobs = await query("SELECT required_skills FROM jobs WHERE status='open'")
    gap_count: dict = {}
    for j in jobs:
        rs = j.get("required_skills")
        if isinstance(rs, str): rs = json.loads(rs)
        for s in (rs or []):
            k = s.lower()
            if k not in my_skills:
                gap_count[k] = gap_count.get(k, 0) + 1

    top_gaps = [s for s, _ in sorted(gap_count.items(), key=lambda x: -x[1])]
    courses = await query("SELECT * FROM courses")
    scored = []
    for c in courses:
        st = c.get("skill_tags")
        if isinstance(st, str): st = json.loads(st)
        tags = [t.lower() for t in (st or [])]
        rel = sum(1 for t in tags if t in top_gaps)
        if rel > 0:
            scored.append({**_course_out(c), "relevance": rel})
    scored.sort(key=lambda x: -x["relevance"])
    return {"topSkillGaps": top_gaps[:6], "recommendedCourses": scored}


@router.get("")
async def list_courses(user: dict = Depends(auth_required)):
    rows = await query("SELECT * FROM courses ORDER BY created_at DESC")
    return [_course_out(r) for r in rows]


@router.post("", status_code=201)
async def create_course(body: dict, user: dict = Depends(require_role("trainer", "admin"))):
    if not body.get("title"): raise HTTPException(400, detail="title is required")
    row = await execute_returning(
        "INSERT INTO courses (trainer_id,title,provider,skill_tags,duration_weeks,level,rating) VALUES ($1,$2,$3,$4,$5,$6,4.5) RETURNING id",
        [user["id"], body["title"], body.get("provider") or user.get("name"),
         json.dumps(body.get("skill_tags") or []),
         body.get("duration_weeks", 4), body.get("level", "Beginner")],
    )
    c = await query_one("SELECT * FROM courses WHERE id=$1", [row["id"]])
    return _course_out(c)


@router.post("/{cid}/enroll", status_code=201)
async def enroll(cid: int, body: dict = None, user: dict = Depends(require_role("candidate"))):
    course = await query_one("SELECT * FROM courses WHERE id=$1", [cid])
    if not course: raise HTTPException(404, detail="Course not found")
    existing = await query_one("SELECT id FROM enrollments WHERE course_id=$1 AND candidate_id=$2", [cid, user["id"]])
    if existing: raise HTTPException(409, detail="Already enrolled")
    batch_id = (body or {}).get("batch_id")
    await execute("INSERT INTO enrollments (course_id, candidate_id, batch_id) VALUES ($1,$2,$3)", [cid, user["id"], batch_id])
    if batch_id:
        be = await query_one("SELECT id FROM batch_enrollments WHERE batch_id=$1 AND candidate_id=$2", [batch_id, user["id"]])
        if not be:
            await execute("INSERT INTO batch_enrollments (batch_id, candidate_id) VALUES ($1,$2)", [batch_id, user["id"]])
    return {"ok": True}
